src/pm.rs

```diff
@@ -1,21 +1,58 @@
 use std::{
     collections::{HashMap, HashSet},
     fs,
-    io::Write,
+    io::{ Read, Write },
 };
 
 use crate::network::{TopicManifest, TopicManifests};
 use crate::parser::list_installed;
-use anyhow::Result;
+use anyhow::{ format_err, Result };
 use serde::{Deserialize, Serialize};
-use serde_json::{from_slice, to_string};
+use serde_json::{from_slice, to_string, Value as jsonValue};
+use serde_yaml::Value as yamlValue;
+use lazy_static::lazy_static;
 
 const SOURCE_HEADER: &[u8] = b"# Generated by AOSC Topic Manager. DO NOT EDIT THIS FILE!\n";
 const SOURCE_PATH: &str = "/etc/apt/sources.list.d/atm.list";
 const STATE_PATH: &str = "/var/lib/atm/state";
 const STATE_DIR: &str = "/var/lib/atm/";
 const DPKG_STATE: &str = "/var/lib/dpkg/status";
-const MIRROR_URL: &str = "https://repo.aosc.io/debs";
+
+fn get_mirror_url() -> Result<String> {
+    let mut status_file = fs::File::open("/var/lib/apt/gen/status.json")?;
+    let mut status_data = String::new();
+    status_file.read_to_string(&mut status_data)?;
+
+    let status_v: jsonValue = serde_json::from_str(&status_data)?;
+    let mirror: String = match &status_v["mirror"][0] {
+        jsonValue::String(m) => String::from(m),
+        _ => {
+            return Err(format_err!("Malformed apt-gen-list data."));
+        }
+    };
+
+    let mut mirrors_file = fs::File::open("/usr/share/distro-repository-data/mirrors.yml")?;
+    let mut mirrors_data = String::new();
+    mirrors_file.read_to_string(&mut mirrors_data)?;
+
+    let mirrors_v: yamlValue = serde_yaml::from_str(&mirrors_data)?;
+    let mirror_url = match &mirrors_v[&mirror]["url"] {
+        yamlValue::String(u) => String::from(u),
+        _ => {
+            return Err(format_err!("Malformed mirrors data."));
+        }
+    };
+
+    Ok(mirror_url)
+}
+
+lazy_static! {
+    static ref MIRROR_URL: String = format!(
+        "{}{}",
+        get_mirror_url().unwrap_or(String::from("https://repo.aosc.io/")),
+        "debs"
+    );
+}
```

Comment:
Is this function the same as that in https://github.com/AOSC-Dev/atm/pull/1/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR80-R115?

Why is this duplicated?

Reply:
Thus do I need to write a separate file to be imported by these two files?

Reply:
> Thus do I need to write a separate file to be imported by these two files?

No. You need to mark your function as either `pub` or `pub(crate)` and then you can import it.

Reply:
Do I need to create a new `mod` or simply mark a `pub` ?

Reply:
You don't need a new module. Just mark the function as public.